File: src/config/GameConfig.js
// GameConfig.js
// Central tuning for the game. Per-run difficulty (which obstacles may spawn and
// how thickly) lives in modes/RunProfile.js and config/JourneyConfig.js; what's
// left here is what every run shares.
// Changes:
// - Open Space Deep space milestone at 12,500 KM (Day 33 belt), not 10,000.
// - Camera reseat knobs (all JS platforms): after 5s below the
//   ideal seat, an 8s ease-in-out creeps the ship back.
// - Soft sparkle magnet: radius 4.25× ship + magnetPull 0.15 so near-miss
//   diamonds ease in more readily; collect still requires contact.
// - Fuel drainPerKm 0.00025 ≈ 4000 KM full tank (playtest: 2700 still too
//   stressful when the next sparkle is far).
// - Added `fuel`: depleting 0–1 tank (distance drain, diamond refill, dying
//   coast). Sparkles collected during the coast cancel dying until the hull stops.
// - Arc banks: arcDuration 820ms for a taller closed swoosh (linear full-π
//   path + 0.55 mid-arc vertical boost in Spacecraft).
// - Zigzag angle/speed tunables are the default flight style (see flightStyle.js).
// - Camera catch-up tightened so direction changes don't feel mushy under a
//   lazy follower.
// - Camera: slightly slower floor cruise than the ship + stronger catch-up so
//   the follower accelerates when the ship leads high (smooth endless-runner feel).
// - Dropped the dead `obstacles.types` unlock table. Nothing read it, and it had
//   drifted from the copy ObstacleManager was actually using; the schedule now
//   lives once, in `OPEN_WORLD_UNLOCKS` (modes/RunProfile.js).
// - Added style-swoosh near-miss tuning (`styleSwoosh`) and `points.perSwoosh`
//   for threading a narrow gap between two obstacles.
// - Added `points`: smash / swoosh style scores (sparkles refill fuel instead).

export const GameConfig = {
    // Survival fuel — separate from KM and from style `points`. Live only once
    // collectibles are enabled for the run (see Game.isFuelLive).
    fuel: {
        max: 1,
        start: 1,
        // HUD KM accrues ~90–110 per second of flight at any playfield size
        // (see kmDelta). Full tank ≈ 4000 KM (~35–40s / ~8–10 sparkle windows)
        // before empty if you take nothing.
        drainPerKm: 0.00025,
        refillPerCollectible: 0.45, // clamp to max; no overfill (~half a tank)
        dyingDurationMs: 900, // coast; salvage sparkles until the hull stops
        lowThreshold: 0.28,
        voiceLowThreshold: 0.20, // NAV warning; HUD pulse stays on lowThreshold
        // Soft magnet assist — sparkles ease toward the ship when close; collect
        // still needs circle overlap. Radius is × spacecraft.radius.
        magnetRadiusScale: 4.25,
        magnetPull: 0.15, // ease per 60fps tick (× tickScale × proximity falloff)
    },
    // Style points — smash / swoosh only. Not a survival meter; not Journey star 2.
    points: {
        perAsteroid: 1,     // destroying one asteroid (with the shield)
        perCollectible: 10, // unused (sparkles refill fuel); kept for reference
        perSwoosh: 15,      // style points for a narrow twin-obstacle near-miss
    },
    // Near-miss "swoosh" — ship squeezed tightly between a left + right obstacle.
    // Clearances are multiples of spacecraft.radius.
    styleSwoosh: {
        maxClearance: 1.65,   // each side must be closer than this
        maxPairYDelta: 3.2,   // left/right obstacles must share a similar Y
        yBand: 1.35,          // ship must be within this of the pair's mid-Y
        cooldownMs: 420,      // min time between swoosh awards
    },
    spacecraft: {
        radius: 1,
        speed: 0.08,
        arcRadius: 0.2,
        // Wall-clock ms for a full closed arc (taller swoosh; linear π in Spacecraft).
        arcDuration: 820,
        // Zigzag (default flight): degrees from straight up (higher = flatter lean).
        zigzagAngleDeg: 52,
        // Along-path speed vs arc cruise (zigzag wants to feel quicker).
        zigzagSpeedScale: 1.45,
        boostSpeed: 0.03,
        trailDotSize: 0.2,
        trailSpacing: 10
    },
    camera: {
        // Floor cruise (× height). Catch-up accelerates above this when the ship
        // leads too high; ship speed is separate (spacecraft.speed).
        speed: 0.07,
        // How hard the camera pulls once the ship drifts past the deadzone.
        // Gentle, so re-seating glides in rather than tugging on turns.
        interpolation: 0.18,
        deceleration: 2000,
        // Velocity ease — lower = snappier catch-up, higher = lazier.
        smoothingFactor: 0.78,
        // Deadzone (× height) around the ideal seat where the camera just follows
        // the ship's speed and applies no re-seating pull — so steady flight is
        // pure smooth scroll and the camera only re-addresses on real drift.
        deadzone: 0.16,
        // After 5s below the ideal seat, ease the leftover gap closed over
        // reseatDuration so a wormhole or black-hole dip does not leave the
        // craft permanently low (all JS platforms: web + Capacitor).
        reseatDelay: 5,
        reseatSlack: 0.03, // × height — ignore tiny dips
        reseatDuration: 8, // seconds to ease the leftover gap closed
        reseatTrack: 0.015, // how tightly we follow the easing target
    },
    obstacles: {
        minSize: 2.5,
        maxSize: 6.25,
        verticalSpacing: 0.35,
        rotationRange: [-0.02, 0.02],
        // Historical Open World cluster scale. Open Space belt density now
        // lives in OpenSpaceWeather.OPEN_SPACE_BELT (KM lerp). Journey ignores this.
        scaling: {
            startDensity: 0.7,
            maxDensity: 1.5,
            rampUpDistance: 10000
        }
    },
    // iOS already awards KM as Δy × (800 / playfieldHeight) × (100/60) so a
    // tall phone and a short canvas stay in lockstep. Web used raw CSS pixels,
    // so filling the desktop viewport (~1080px vs ~750) burned a full tank by
    // ~2000 KM. Same helper here — fuel and HUD KM ignore screen size.
    kmPerPixel: 100 / 60,
    kmReferenceHeight: 800,
    kmDelta(dy, playfieldHeight) {
        const scale = this.kmReferenceHeight / Math.max(playfieldHeight, 1);
        return Math.abs(dy) * scale * this.kmPerPixel;
    },
    milestones: [
        { score: 1000, message: "Breaking atmosphere..." },
        { score: 2000, message: "Warning: Complex asteroids detected..." },
        { score: 5000, message: "Caution: Asteroid belts ahead..." },
        { score: 12500, message: "Deep space detected..." },
        { score: 25000, message: "Unknown signals ahead..." },
        { score: 50000, message: "Approaching the void..." }
    ]
}; 

File: ios-native/SpaceSwoosh/Core/GameConfig.swift
// GameConfig.swift
// Changes: driftLaneSlots (4 currents × 7 Android shear lines). Android
// catch-up camera knobs + mobile baseUnit (min(w/45, h/75)). feelSpeed 1.0
// so native iOS cruise matches Android/web snappy tick. Fuel.dyingStopSpeed
// 0.01 so fuel-out waits until the hull stops. Fuel.voiceLowThreshold 0.20
// for NAV low-fuel lines (HUD stays 0.28). Dropped 1000 KM "Breaking
// atmosphere..." HUD line.

import Foundation
import CoreGraphics

enum GameConfig {
    static let simDt: CGFloat = 1.0 / 60.0
    /// JS `snappyHz`: ship travel uses `speed * (1/60) * (dt * 120)` so one
    /// 60 Hz step covers the same distance as Android's snappy pacing.
    static let snappyHz: CGFloat = 120
    /// iOS feel knob — 1.0 matches Android/web snappy travel (was 0.90).
    static let feelSpeed: CGFloat = 1.0
    static func motionTickScale(dt: CGFloat) -> CGFloat { dt * snappyHz * feelSpeed }
    /// JS: abs(Δcamera.y) * (100/60) in CSS pixels. Scale to this height so
    /// a tall iPhone and a short Android CSS canvas award KM at the same pace.
    static let kmPerPixel: CGFloat = 100.0 / 60.0
    static let kmReferenceHeight: CGFloat = 800

    static func kmDelta(dy: CGFloat, playfieldHeight: CGFloat) -> CGFloat {
        let scale = kmReferenceHeight / max(playfieldHeight, 1)
        return abs(dy) * scale * kmPerPixel
    }

    enum Spacecraft {
        static let radiusUnits: CGFloat = 1
        static let speed: CGFloat = 0.08
        static let zigzagAngleDeg: CGFloat = 52
        static let zigzagSpeedScale: CGFloat = 1.45
        static let arcRadius: CGFloat = 0.2
        static let arcDurationMs: CGFloat = 820
        static let arcVerticalBoost: CGFloat = 0.55
        static let trailMaxPoints: Int = 80
        static let trailFadePerTick: CGFloat = 1.0 / 180.0
        static let trailDotSize: CGFloat = 0.2
        static let trailSpacing: CGFloat = 10
        static let maxBank: CGFloat = 0.96
        static let tailOffset: CGFloat = 0.6
        /// JS `BANK_SMOOTHING` — lean eases by elapsed time, not a snap.
        static let bankSmoothing: CGFloat = 0.34
    }

    enum Flicker {
        static let trailWidthScale: CGFloat = 0.6
        static let wallJellyMs: CGFloat = 420
        static let boopCooldownMs: CGFloat = 180
        static let shieldHitboxScale: CGFloat = 1.5
        static let hullDrawPad: CGFloat = 3.2
        /// iOS feel — pickup / portal / clear flyout. Android remains 5s.
        static let shieldSeconds: CGFloat = 4.0
    }

    enum Camera {
        /// Floor cruise (× height). Ship speed is separate (Spacecraft.speed).
        static let speed: CGFloat = 0.07
        static let interpolation: CGFloat = 0.18
        static let smoothingFactor: CGFloat = 0.78
        static let deadzone: CGFloat = 0.16
        static let reseatDelay: CGFloat = 5
        static let reseatSlack: CGFloat = 0.03
        static let reseatDuration: CGFloat = 8
        static let reseatTrack: CGFloat = 0.015
        /// From bottom. Android `idealOffset` 0.75 from top.
        static let idealSeat: CGFloat = 0.25
        /// Android `camera.deceleration` — game-over scroll fade.
        static let decelerationMs: CGFloat = 2000
    }

    enum Playfield {
        /// Full device; no letterbox.
        static let fillsDevice = true
        /// Android mobile `baseUnit`: min(width/45, height/75).
        static func baseUnit(width: CGFloat, height: CGFloat) -> CGFloat {
            min(width / 45, height / 75)
        }
    }

    enum Fuel {
        static let max: CGFloat = 1
        static let start: CGFloat = 1
        static let drainPerKm: CGFloat = 0.00025
        static let refillPerCollectible: CGFloat = 0.45
        static let dyingDurationMs: CGFloat = 900
        /// JS MIN_HEADING_SPEED — fuel-out fail waits until displacement is noise.
        static let dyingStopSpeed: CGFloat = 0.01
        static let lowThreshold: CGFloat = 0.28
        static let voiceLowThreshold: CGFloat = 0.20
        static let magnetRadiusScale: CGFloat = 4.25
        static let magnetPull: CGFloat = 0.15
    }

    enum Points {
        static let perAsteroid = 1
        static let perSwoosh = 15
    }

    enum StyleSwoosh {
        static let maxClearance: CGFloat = 1.65
        static let maxPairYDelta: CGFloat = 3.2
        static let yBand: CGFloat = 1.35
        static let cooldownMs: CGFloat = 420
    }

    enum Profile {
        static let shieldsFromScore: CGFloat = 500
        static let collectiblesFromScore: CGFloat = 100
        static let wallBoostsFromScore: CGFloat = 12000
        static let simpleChance: CGFloat = 0.65
        static let maxRowSpawns = 3
        static let maxClusterCount = 4
        static let sparkleFirstWait: CGFloat = 3.2
        static let sparkleMin: CGFloat = 2.6
        static let sparkleSpan: CGFloat = 2.6
        static let shieldInterval: CGFloat = 5
        static let wallBoostInterval: CGFloat = 22
    }

    enum Obstacles {
        /// Set-piece size band (JS `config.obstacles`). Simple clusters use 0.9…1.4.
        static let minSizeUnits: CGFloat = 2.5
        static let maxSizeUnits: CGFloat = 6.25
        static let scaling = (startDensity: CGFloat(0.7), maxDensity: CGFloat(1.5), rampUpDistance: CGFloat(10000))
    }

    enum Unlocks {
        static let table: [(type: String, score: CGFloat, message: String)] = [
            ("simple", 0, "Watch out for asteroids!"),
            ("sideBarrier", 1000, "Warning: Side barriers detected!"),
            ("complex", 1000, "Warning: Asteroids with orbiting debris detected!"),
            ("moving", 2000, "Caution: Moving asteroids detected!"),
            ("shooting", 3000, "Warning: Hostile asteroids detected!"),
            ("driftCurrent", 3500, "Crosswinds detected — lateral currents ahead!"),
            ("pulsating", 4000, "Warning: Unstable asteroids ahead!"),
            ("phase", 4500, "Square blooms detected — they spring open and shove!"),
            ("wormhole", 5000, "Spatial anomalies detected!"),
            ("repulsor", 5500, "Repulsor fields detected — they push, not pull!"),
            ("blackhole", 6000, "Gravitational anomalies detected!"),
            ("sweepGate", 7000, "Sweep lines ahead — timed corridors!"),
        ]
    }

    enum Milestones {
        static let table: [(score: CGFloat, message: String)] = [
            (2000, "Warning: Complex asteroids detected..."),
            (5000, "Caution: Asteroid belts ahead..."),
            (12500, "Deep space detected..."),
            (25000, "Unknown signals ahead..."),
            (50000, "Approaching the void..."),
        ]
        static let teachKm: CGFloat = 80
        static let atmosphereKm: CGFloat = 200
    }

    enum Stress {
        static let obstacleSlots = 64
        static let extraPartSlots = 48
        /// 4 on-screen drift currents × 7 shear lines (Android DriftCurrent).
        static let driftLaneSlots = 28
        static let sparkleSlots = 24
        static let glowSlots = 16
        static let pickupSlots = 16
        static let recycleLeadScreens: CGFloat = 2.4
        static let recycleBehindScreens: CGFloat = 0.35
    }
}
